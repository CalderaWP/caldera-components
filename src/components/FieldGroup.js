import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup,FormControl,ControlLabel,HelpBlock } from 'react-bootstrap';
export function FieldGroup({ id, label, help, ...props }) {
	return (
		<FormGroup controlId={id}>
			<ControlLabel
				controlId={id}
			>
				{label}
			</ControlLabel>
			<FormControl {...props} />
			{help && <HelpBlock>{help}</HelpBlock>}
		</FormGroup>
	);
}

FieldGroup.propTypes = {
	id: PropTypes.string.isRequired,
	props: PropTypes.object,
	help: PropTypes.string,
	label: PropTypes.string.isRequired
};