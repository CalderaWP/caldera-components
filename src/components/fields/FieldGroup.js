import React from 'react';
import classNames from 'classnames';
import {addOrRemoveFromArray, ariaDescribedbyAttr} from './util';
import {
	fieldGroupPropTypes,
} from './propTypes';

import {FieldInner} from './FieldInner';
import {RenderGroup} from '../RenderGroup';
import {Message} from './messages/Message';
import {messageObjectFactory} from './messages/messageObjectFactory';

/**
 * Represents one configField -- wrapper, label and input.
 *
 * @param props
 * @return {*}
 * @constructor
 */
export const FieldGroup = (props) => {

	/**
	 * Creates the id attribute
	 * @return {String}
	 */
	function idAttrForHelpElement() {
		return ariaDescribedbyAttr(props.id, props.help);
	}

	function checkboxChangeHandler(option,values) {
		if (!Array.isArray(values)) {
			if ('string' === typeof values || 'number' === typeof  values) {
				values = [values];
			} else {
				values = [];

			}
		}
		values = addOrRemoveFromArray(option.value, values);
		props.onValueChange(values);
	}

	/**
	 * Based on type, render the field.
	 *
	 * @return {*}
	 */
	function fieldInner(fieldProps){
		const message = 'object' === typeof  fieldProps.message
			? messageObjectFactory(fieldProps.message)
			: messageObjectFactory({message:null, error: false });
		function getFieldClassName(conditionalClassNames) {
			return classNames(
				{
					required: fieldProps.isRequired,
					...conditionalClassNames
				}
			);
		}

		/**
		 * Is this option the checked value?
		 * @param {object}option
		 * @param {String|number|Array} currentValue
		 * @return {boolean}
		 */
		function isCheckedOption( option, currentValue){
			if( Array.isArray( currentValue ) ){
				return currentValue.includes(option.value);
			}
			return option.value === currentValue;
		}

		//Fieldsets are rendered recursively.
		if( 'fieldset' === fieldProps.type ){
			return (
				<fieldset
					id={props.id}
				>
					<legend>
						{props.label}
					</legend>
					{props.options.map(option =>  {


						//Call this same function, as a regular checkbox
						return (
							<React.Fragment
								key={ `${props.id}-${option.value}` }
							>
								{fieldInner({
									type: 'input',
									inputType: 'checkbox',
									id: `${props.id}-${option.value}`,
									value: isCheckedOption(option,props.value),
									fieldClassName: getFieldClassName({
										'fieldset-checkbox' : true,
									}),
									label: option.label,
									disabled:props.disabled,
									onValueChange: () => {
										checkboxChangeHandler(option,props.values);
									}
								})}
							</React.Fragment>
						);
					})}
				</fieldset>
			);
		}else if( 'hidden' === fieldProps.inputType ){
			return (
				FieldInner(
					{
						type: props.type,
						id: props.id,
						fieldClassName: getFieldClassName(),
						value:fieldProps.value,
						onValueChange:fieldProps.onValueChange,
						inputType: fieldProps.inputType,
						disabled:props.disabled,
					}
				)
			);
		}else{
			return (
				<div
					className={RenderGroup.classNames.fieldWrapper}
				>
					<label
						htmlFor={fieldProps.id}
					>
						{fieldProps.label}
					</label>
					<Message
						message={message}
					/>
					{FieldInner(
						{
							type: fieldProps.type,
							id: fieldProps.id,
							fieldClassName: getFieldClassName({
								'block-input': fieldProps.isBlockInput
							}),
							help:fieldProps.help,
							value:fieldProps.value,
							onValueChange:fieldProps.onValueChange,
							inputType: fieldProps.inputType,
							options: fieldProps.options,
							disabled:props.disabled,
						}
					)}
					{fieldProps.help &&
					<p
						id={`${idAttrForHelpElement()}`}
						className={'description'}
					>
						{fieldProps.help}
					</p>
					}
				</div>
			);
		}

	}

	/**
	 * Render field group
	 */
	return fieldInner(props);
};




FieldGroup.propTypes = fieldGroupPropTypes;

FieldGroup.defaultProps = {
	isBlockInput: true,
	isRequired: false,
	help: ''
};