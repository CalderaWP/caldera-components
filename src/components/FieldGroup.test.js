import {FieldGroup} from "./FieldGroup";
import renderer from 'react-test-renderer';
import React from 'react';

describe( 'Field group component', () => {
	it('Works with help text',() => {
		const fieldGroup = renderer.create(
			<FieldGroup
				id={'control-22'}
				label={'Who'}
				help={'Who to say hi to'}
			/>
		);
		expect( fieldGroup.toJSON() ).toMatchSnapshot();
	});

	it('Help text is optional',() => {
		const fieldGroup = renderer.create(
			<FieldGroup
				id={'control-22'}
				label={'Who'}
			/>
		);
		expect( fieldGroup.toJSON() ).toMatchSnapshot();
	});
});