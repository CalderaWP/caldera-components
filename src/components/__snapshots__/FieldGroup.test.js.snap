// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field group component Help text is optional 1`] = `
<div
  className="form-group"
>
  <label
    className="control-label"
    controlId="control-22"
    htmlFor="control-22"
  >
    Who
  </label>
  <input
    className="form-control"
    id="control-22"
  />
</div>
`;

exports[`Field group component Works with help text 1`] = `
<div
  className="form-group"
>
  <label
    className="control-label"
    controlId="control-22"
    htmlFor="control-22"
  >
    Who
  </label>
  <input
    className="form-control"
    id="control-22"
  />
  <span
    className="help-block"
  >
    Who to say hi to
  </span>
</div>
`;
